parlai/agents/rag/retrievers.py

```diff
@@ -0,0 +1,1217 @@
+        encoder_type: str = 'query',
+    ):
+        # Override options
+        config: BertConfig = BertConfig.from_pretrained('bert-base-uncased')
+        pretrained_path = modelzoo_path(opt['datapath'], pretrained_path)
+        if not os.path.exists(pretrained_path):
+            # when initializing from parlai rag models, the pretrained path
+            # may not longer exist. This is fine if we've alread trained
+            # the model.
+            assert dpr_model == 'bert_from_parlai_rag'
+            logging.error(f'Pretrained Path does not exist: {pretrained_path}')
+            pretrained_path = modelzoo_path(opt['datapath'], DPR_ZOO_MODEL)
+            dpr_model = 'bert'
+            logging.error(f'Setting to zoo model: {pretrained_path}')
+        enc_opt = {
+            "n_heads": config.num_attention_heads,
+            "n_layers": config.num_hidden_layers,
+            "embedding_size": config.hidden_size,
+            "ffn_size": config.intermediate_size,
+            "dropout": config.hidden_dropout_prob,
+            "attention_dropout": config.attention_probs_dropout_prob,
+            "activation": config.hidden_act,
+            "variant": 'xlm',
+            "reduction_type": 'first',
+            "n_positions": config.max_position_embeddings,
+            "n_segments": config.type_vocab_size,
+        }
+        embedding = torch.nn.Embedding(
+            config.vocab_size, config.hidden_size, padding_idx=config.pad_token_id
+        )
+        super().__init__(
+            enc_opt,
+            vocabulary_size=config.vocab_size,
+            padding_idx=config.pad_token_id,
+            embedding=embedding,
+            reduction_type='first',
+        )
+
+        self._load_state(dpr_model, pretrained_path, encoder_type)
+
+    def _load_state(self, dpr_model: str, pretrained_path: str, encoder_type: str):
+        """
+        Load pre-trained model states.
+
+        :param dpr_model:
+            which dpr model type we're using
+        :param pretrained_path:
+            path to pretrained model
+        :param encoder_type:
+            whether this is a query or document encoder
+        """
+        if dpr_model == 'bert':
+            state_dict = BertConversionUtils.load_bert_state(
+                self.state_dict(),
+                pretrained_dpr_path=pretrained_path,
+                encoder_type=encoder_type,
+            )
+            self.load_state_dict(state_dict)
+        elif dpr_model == 'bert_from_parlai_rag':
+            state_dict = torch.load(pretrained_path, map_location='cpu')["model"]
+            key = f"{encoder_type}_encoder."
+            state_dict = {
+                k.split(key)[-1]: v for k, v in state_dict.items() if key in k
+            }
+            self.load_state_dict(state_dict)
+
+
+class RagDprQueryEncoder(RagDprEncoder):
+    """
+    Query Encoder for DPR.
+    """
+
+    def __init__(self, *args, **kwargs):
+        kwargs['encoder_type'] = 'query'
```

Comment:
Instead of this, what if we just override `_load_state` in `RagDprQueryEncoder` and `RagDprDocumentEncoder`?

Intuition tells me it's less error-prone that way, but I'm struggling to come up with a precise explanation/example for why right now.

Reply:
ehh i think overriding `_load_state` would come at the cost of lots of redundant code, which could be solved by setting like `self.encoder_type`, which could be clean but just as error-prone as this is haha

Reply:
Ah yeah I see. Because query vs document encoder is a separate decision to choosing a `dpr_model`. Missed that the first time around.